pkg/network/common/egress_dns.go

```diff
@@ -5,89 +5,110 @@ import (
 	"sync"
 	"time"
 
+	networkv1 "github.com/openshift/api/network/v1"
 	ktypes "k8s.io/apimachinery/pkg/types"
 	utilruntime "k8s.io/apimachinery/pkg/util/runtime"
-
-	networkv1 "github.com/openshift/api/network/v1"
+	"k8s.io/apimachinery/pkg/util/sets"
+	"k8s.io/klog"
 )
 
-type EgressDNSUpdate struct {
+type PolicyUpdate struct {
 	UID       ktypes.UID
 	Namespace string
 }
 
+type PolicyUpdates []PolicyUpdate
+
 type EgressDNS struct {
 	// Protects pdMap/namespaces operations
 	lock sync.Mutex
-	// Holds Egress DNS entries for each policy
-	pdMap map[ktypes.UID]*DNS
+	// holds DNS entries globally
+	dns *DNS
+	// this map holds which DNS names are in what policy objects
+	dnsNamesToPolicies map[string]sets.String
 	// Maintain namespaces for each policy to avoid querying etcd in syncEgressDNSPolicyRules()
 	namespaces map[ktypes.UID]string
 
 	// Report change when Add operation is done
 	added chan bool
 
 	// Report changes when there are dns updates
-	Updates chan EgressDNSUpdate
+	Updates chan PolicyUpdates
 }
 
-func NewEgressDNS() *EgressDNS {
-	return &EgressDNS{
-		pdMap:      map[ktypes.UID]*DNS{},
-		namespaces: map[ktypes.UID]string{},
-		added:      make(chan bool),
-		Updates:    make(chan EgressDNSUpdate),
-	}
-}
-
-func (e *EgressDNS) Add(policy networkv1.EgressNetworkPolicy) {
+func NewEgressDNS() (*EgressDNS, error) {
 	dnsInfo, err := NewDNS("/etc/resolv.conf")
 	if err != nil {
 		utilruntime.HandleError(err)
+		return nil, err
 	}
+	return &EgressDNS{
+		dns:                dnsInfo,
+		dnsNamesToPolicies: map[string]sets.String{},
+		namespaces:         map[ktypes.UID]string{},
+		added:              make(chan bool),
+		Updates:            make(chan PolicyUpdates),
+	}, nil
+}
+
+func (e *EgressDNS) Add(policy networkv1.EgressNetworkPolicy) {
+	e.lock.Lock()
+	defer e.lock.Unlock()
 
 	for _, rule := range policy.Spec.Egress {
 		if len(rule.To.DNSName) > 0 {
-			if err := dnsInfo.Add(rule.To.DNSName); err != nil {
-				utilruntime.HandleError(err)
+			if uids, exists := e.dnsNamesToPolicies[rule.To.DNSName]; !exists {
+				e.dnsNamesToPolicies[rule.To.DNSName] = sets.NewString(string(policy.UID))
+				//only call Add if the dnsName doesn't exist in the dnsNamesToPolicies
+				if err := e.dns.Add(rule.To.DNSName); err != nil {
+					utilruntime.HandleError(err)
+				}
+
+			} else {
+				e.dnsNamesToPolicies[rule.To.DNSName] = uids.Insert(string(policy.UID))
```

Comment:
You don't need to re-set `e.dnsNamesToPolicies[rule.To.DNSName]`. `uids` will still point to the same object.

Reply:
Agreed, this is to make sure that if someone changes the value type back to a slice, the code will still work as is. Of course the `Insert` would have to change to `append()` but then it's easy to forget assigning the result of the `append` as the value type of the map entry.